File: src/eopx/format/shamir.py
"""Shamir secret sharing over GF(2^8) using the AES irreducible polynomial.

The field is :math:`F_{256} = F_2[X] / (X^8 + X^4 + X^3 + X + 1)` (Rijndael
polynomial ``0x11B``). Each byte of the secret is split independently:

* Random polynomial ``p(X) = a_0 + a_1 X + ... + a_{k-1} X^{k-1}`` over
  GF(2^8), with ``a_0`` set to the secret byte and ``a_1..a_{k-1}``
  drawn from :mod:`secrets`.
* The *i*-th share is ``p(i)`` evaluated at ``x = i`` for
  ``i in {1, 2, ..., n}`` (``x = 0`` would leak the secret).
* Reconstruction interpolates ``p(0)`` from any ``k`` shares via the
  Lagrange formula.

Security
--------
* Any ``k-1`` shares reveal NOTHING about the secret beyond its length.
* Shares are **fixed-length** equal to the secret's length: this leaks
  the secret length but no content.
* This implementation does NOT include integrity protection — that is
  the job of the surrounding :mod:`eopx.format.visual_sharding` layer
  (AEAD via ChaCha20-Poly1305).
"""

from __future__ import annotations

import secrets
from typing import Iterable, List, Sequence, Tuple

# Rijndael irreducible polynomial: x^8 + x^4 + x^3 + x + 1
_RIJNDAEL = 0x11B

# Precomputed log/antilog tables for fast GF(2^8) multiply / inverse.
_EXP: List[int] = [0] * 512
_LOG: List[int] = [0] * 256


def _init_tables() -> None:
    """Build log / antilog tables using generator 0x03."""
    g = 0x03
    x = 1
    for i in range(255):
        _EXP[i] = x
        _LOG[x] = i
        x = _mul_naive(x, g)
        # _mul_naive is only used here during bootstrap
    for i in range(255, 512):
        _EXP[i] = _EXP[i - 255]


def _mul_naive(a: int, b: int) -> int:
    """Schoolbook GF(2^8) multiplication, used only for table bootstrap."""
    p = 0
    for _ in range(8):
        if b & 1:
            p ^= a
        hi = a & 0x80
        a = (a << 1) & 0xFF
        if hi:
            a ^= 0x1B
        b >>= 1
    return p


def gf_mul(a: int, b: int) -> int:
    """Multiply two GF(2^8) elements (returns ``a * b`` in the field)."""
    if a == 0 or b == 0:
        return 0
    return _EXP[_LOG[a] + _LOG[b]]


def gf_inv(a: int) -> int:
    """Multiplicative inverse in GF(2^8). Raises on ``a == 0``."""
    if a == 0:
        raise ZeroDivisionError("0 has no inverse in GF(2^8)")
    return _EXP[255 - _LOG[a]]


def gf_div(a: int, b: int) -> int:
    """Divide ``a / b`` in GF(2^8)."""
    if a == 0:
        return 0
    if b == 0:
        raise ZeroDivisionError("division by zero in GF(2^8)")
    return _EXP[(_LOG[a] - _LOG[b]) % 255]


_init_tables()


# ---------------------------------------------------------------------------
# Polynomial evaluation
# ---------------------------------------------------------------------------

def _eval_poly(coeffs: Sequence[int], x: int) -> int:
    """Horner evaluation of a polynomial over GF(2^8) at ``x``."""
    acc = 0
    for c in reversed(coeffs):
        acc = gf_mul(acc, x) ^ c
    return acc


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

Share = Tuple[int, bytes]


def shamir_split(secret: bytes, k: int, n: int) -> List[Share]:
    """Split ``secret`` into ``n`` shares; any ``k`` shares reconstruct it.

    Parameters
    ----------
    secret:
        Arbitrary bytes (at least 1 byte).
    k:
        Threshold. Must satisfy ``1 <= k <= n``.
    n:
        Total number of shares. Must satisfy ``k <= n <= 255``.

    Returns
    -------
    list of (index, share_bytes)
        Indices run from 1 to n (never 0). Share bytes have the same
        length as the secret.
    """
    if not isinstance(secret, (bytes, bytearray)) or len(secret) == 0:
        raise ValueError("secret must be non-empty bytes")
    if not (1 <= k <= n):
        raise ValueError(f"need 1 <= k <= n, got k={k}, n={n}")
    if n > 255:
        raise ValueError("n must be <= 255 for GF(2^8) Shamir")

    secret_bytes = bytes(secret)
    L = len(secret_bytes)
    shares: List[bytearray] = [bytearray(L) for _ in range(n)]

    # For each byte position, build a fresh random polynomial.
    for pos in range(L):
        coeffs = [secret_bytes[pos]]
        # k-1 random coefficients in GF(2^8). secrets.randbits is CSPRNG.
        for _ in range(k - 1):
            coeffs.append(secrets.randbits(8))
        for i in range(1, n + 1):
            shares[i - 1][pos] = _eval_poly(coeffs, i)

    return [(i + 1, bytes(s)) for i, s in enumerate(shares)]


def shamir_combine(shares: Iterable[Share]) -> bytes:
    """Reconstruct the original secret from at least ``k`` shares.

    Parameters
    ----------
    shares:
        Iterable of ``(index, share_bytes)``. Indices must be distinct
        and in ``1..255``. All share byte strings must have the same
        length.

    Returns
    -------
    bytes
        The recovered secret. If fewer than the true threshold are
        provided, this function still returns *something* — Shamir
        cannot detect "too few shares"; only the caller knows the
        threshold ``k``. Pair with an AEAD check upstream to confirm
        success.
    """
    share_list = list(shares)
    if not share_list:
        raise ValueError("at least one share is required")
    L = len(share_list[0][1])
    indices = [i for i, _ in share_list]
    if len(set(indices)) != len(indices):
        raise ValueError("share indices must be distinct")
    if any(i < 1 or i > 255 for i in indices):
        raise ValueError("share indices must lie in 1..255")
    if any(len(s) != L for _, s in share_list):
        raise ValueError("all shares must have equal length")

    out = bytearray(L)
    # Precompute Lagrange basis values at x = 0:
    #   L_i(0) = prod_{j != i} x_j / (x_i XOR x_j)
    coeffs0: List[int] = []
    for i_idx, xi in enumerate(indices):
        num = 1
        den = 1
        for j_idx, xj in enumerate(indices):
            if j_idx == i_idx:
                continue
            num = gf_mul(num, xj)
            den = gf_mul(den, xi ^ xj)
        coeffs0.append(gf_div(num, den))

    for pos in range(L):
        acc = 0
        for (_, share_bytes), c0 in zip(share_list, coeffs0):
            acc ^= gf_mul(share_bytes[pos], c0)
        out[pos] = acc
    return bytes(out)

File: pwa/src/lib/shamir.ts
/**
 * Shamir secret sharing over GF(2^8) with the Rijndael polynomial 0x11B.
 *
 * Pure TypeScript port of ``eopx.format.shamir`` — same log/antilog
 * tables, same polynomial encoding, same Lagrange reconstruction. The
 * randomness for split() comes from ``crypto.getRandomValues``.
 *
 * Split with ``split(secret, k, n)`` always returns indices ``1..n``;
 * combine with ``combine(shares)`` reconstructs the secret if at least
 * ``k`` distinct shares are supplied. Fewer than ``k`` shares produce
 * garbage — pair this with the AEAD layer in ``recovery.ts`` to detect
 * insufficient input.
 */

export type Share = { index: number; bytes: Uint8Array };

const EXP = new Uint8Array(512);
const LOG = new Uint8Array(256);

function mulNaive(a: number, b: number): number {
  let p = 0;
  for (let i = 0; i < 8; i++) {
    if (b & 1) p ^= a;
    const hi = a & 0x80;
    a = (a << 1) & 0xff;
    if (hi) a ^= 0x1b;
    b >>= 1;
  }
  return p;
}

(function initTables() {
  const g = 0x03;
  let x = 1;
  for (let i = 0; i < 255; i++) {
    EXP[i] = x;
    LOG[x] = i;
    x = mulNaive(x, g);
  }
  for (let i = 255; i < 512; i++) EXP[i] = EXP[i - 255];
})();

export function gfMul(a: number, b: number): number {
  if (a === 0 || b === 0) return 0;
  return EXP[LOG[a] + LOG[b]];
}

export function gfInv(a: number): number {
  if (a === 0) throw new Error("0 has no inverse in GF(2^8)");
  return EXP[255 - LOG[a]];
}

export function gfDiv(a: number, b: number): number {
  if (a === 0) return 0;
  if (b === 0) throw new Error("division by zero in GF(2^8)");
  // (LOG[a] - LOG[b]) mod 255 — handle the negative branch like Python's %
  let d = LOG[a] - LOG[b];
  if (d < 0) d += 255;
  return EXP[d];
}

function evalPoly(coeffs: ArrayLike<number>, x: number): number {
  let acc = 0;
  for (let i = coeffs.length - 1; i >= 0; i--) {
    acc = gfMul(acc, x) ^ coeffs[i];
  }
  return acc & 0xff;
}

/**
 * Split a secret into ``n`` shares; any ``k`` recover it.
 *
 * The optional ``randomCoeff`` lets test code inject deterministic
 * coefficients (used to verify parity against pre-computed Python
 * vectors). Production callers must leave it undefined.
 */
export function split(
  secret: Uint8Array,
  k: number,
  n: number,
  randomCoeff?: (pos: number, j: number) => number,
): Share[] {
  if (!(secret instanceof Uint8Array) || secret.length === 0)
    throw new Error("secret must be non-empty bytes");
  if (!(1 <= k && k <= n)) throw new Error(`need 1 <= k <= n, got k=${k}, n=${n}`);
  if (n > 255) throw new Error("n must be <= 255 for GF(2^8) Shamir");

  const L = secret.length;
  const shares: Uint8Array[] = [];
  for (let i = 0; i < n; i++) shares.push(new Uint8Array(L));

  const randBuf = randomCoeff ? null : new Uint8Array(k - 1);
  for (let pos = 0; pos < L; pos++) {
    if (randBuf) crypto.getRandomValues(randBuf);
    const coeffs = new Uint8Array(k);
    coeffs[0] = secret[pos];
    for (let j = 1; j < k; j++) {
      coeffs[j] = randomCoeff
        ? randomCoeff(pos, j) & 0xff
        : (randBuf as Uint8Array)[j - 1];
    }
    for (let i = 1; i <= n; i++) {
      shares[i - 1][pos] = evalPoly(coeffs, i);
    }
  }
  return shares.map((b, i) => ({ index: i + 1, bytes: b }));
}

/** Lagrange-interpolate the secret from ``shares.length`` shares at x=0. */
export function combine(shares: Share[]): Uint8Array {
  if (shares.length === 0)
    throw new Error("at least one share is required");
  const L = shares[0].bytes.length;
  const indices = shares.map((s) => s.index);
  if (new Set(indices).size !== indices.length)
    throw new Error("share indices must be distinct");
  if (indices.some((i) => i < 1 || i > 255))
    throw new Error("share indices must lie in 1..255");
  if (shares.some((s) => s.bytes.length !== L))
    throw new Error("all shares must have equal length");

  const coeffs0: number[] = [];
  for (let i = 0; i < indices.length; i++) {
    const xi = indices[i];
    let num = 1;
    let den = 1;
    for (let j = 0; j < indices.length; j++) {
      if (j === i) continue;
      const xj = indices[j];
      num = gfMul(num, xj);
      den = gfMul(den, xi ^ xj);
    }
    coeffs0.push(gfDiv(num, den));
  }

  const out = new Uint8Array(L);
  for (let pos = 0; pos < L; pos++) {
    let acc = 0;
    for (let i = 0; i < shares.length; i++) {
      acc ^= gfMul(shares[i].bytes[pos], coeffs0[i]);
    }
    out[pos] = acc;
  }
  return out;
}
